### DISCUSSION TITLE
What is the type of “props.moves” in the React client's board?

### DISCUSSION THREAD

**Discussion by jakehsiao** [boardgameio/boardgame.io#903] [Q&A] [answered]
I know the type of "props.G" is defined by the project, and "props.ctx" is "Ctx", but what is the type of "props.moves"?

`
export function Board(props: {G: IGame, ctx: Ctx, moves: any}) {
 ...
}
`

Currently I am using "any" but this is not a safe way at all.

Also, it's quite appreciated to add some examples and documentations on using boardgame.io with typescript.

**Comment by delucis** [comment, marked as answer]
[Here's a small example using Typescript](https://codesandbox.io/s/boardgameio-typescript-demo-u5uvm), but yes proper TS documentation is still needed.

We don't have strongly typed moves, although it might be theoretically possible to compute them from a game definition, so for now `moves` is

```ts
type moves = Record<
  string,
  (...args: any[]) => void
>;
```

If you follow the example linked above though you shouldn't have to type this manually. If you specifically need the moves type you can use `BoardProps['moves']`.

**Comment by jakehsiao** [comment]
@delucis Thanks, this demo is quite helpful. Maybe it's impossible for typescript now to give a 100% type check.

**Comment by delucis** [comment]
@jakehsiao It’s tricky. If you declare your game as `const game = { ... } as const`, in theory Typescript can read all the moves keys and iterate over them to produce the type of `props.moves` ([here’s an experimental implementation][impl]). However, then you have to type everything in the game definition manually. If you use the `Game` type (or other types like `Move`), like `const game: Game = { ... }`, then Typescript can’t parse specific moves any more as it just sees the generic `Game` type.

I’m not sure the trade-off of typing everything manually is worth it, even though having auto-completion and type safety on the client would be nice.

[impl]: https://github.com/boardgameio/boardgame.io/pull/801#pullrequestreview-487295970

**Comment by drmercer** [comment]
@delucis I believe one way to get the best of both worlds (detailed type information + automatic type checking) would be to use an identity function with a generic type signature, like so:
```ts
function defineGame<Gm extends Game>(game: Gm): Gm {
  return game;
}

// example usage - MyGame will have detailed type information while ensuring it complies with the Game type.
const MyGame = defineGame({
  moves: {
    ...
  },
});
```

Then you could use the experimental type linked in your answer to get the type of `client.moves` based on the type of `MyGame`.

Vue uses this pattern for similar reasons with their `defineComponent` function - see [here](https://v3.vuejs.org/guide/typescript-support.html#defining-vue-components).

**Reply by delucis** [reply]
Interesting idea. Funnily enough, there used to be a `Game()` wrapper function you had to pass your game object to. Now it just gets called internally to simplify the public API.

I’m a bit hesitant about adding things like this only for typing purposes, and this also only partially solves the issue. It’s fairly rare to write out the full game definition in a single object literal — usually moves get broken into separate functions, bits of config get broken up for re-use in various scenarios, etc. I guess you could still extend this logic and have `defineMove`, `defineTurn`, `define…` helpers etc., although this does add quite a bit of boilerplate.

One other stumbling point is you still need to assert `as const`:

```ts
const game = defineGame({ moves: {} } as const);
```

Also this makes it tricky to type `G` because of https://github.com/microsoft/TypeScript/issues/26242. Currently, you can type your game as `Game<G>` and it will type check your `G` in moves and elsewhere. However, because currently there’s no support for inferring only some of the type arguments, we’d need a workaround like:

```ts
function defineGame<G extends any = any>() {
  return function <Gm extends Game<G>>(game: Gm): Gm {
    return game;
  }
}

interface MyGameState {
  score: number;
}

export const MyGame = defineGame<MyGameState>()({ moves: {} } as const);
```
[TS Playground ↗︎](https://www.typescriptlang.org/play?#code/PTAEHEEMFsFMDoAuBnAUAS2gBwPYCdFREBPLWUAbwhnIF9QAzPHaUAcgCMdI8ATAcxrx0ONgG5UGAHaJYeBpADG5ALLEocAMqJIsyqlCgQoSAGtI7cGwA0oYjgCueUILhtkoZDtmpakhg5SiogiUqC8sAzoUrAasAA84KCwAB6yUrwekFLEoAC8JjkAfAAUAJT6hniwiE5hAUEhOGGJrKnpmdRwiUWlrrAAXBDQZUPgrBQGhqDVtXhh-RKGfn6oqbgEoIrNXqBqcfnhkdGxNPH7NNq6sKVlJVTQOABusMhDFPT0kB7bUl5lEiAA)

Works! But is I think maybe more confusing than anything? Admittedly, this could partially be a userland thing. If we expect games to extend `Game` as we currently do, it wouldn’t matter if you used an object that is just `Game` or something like the above returns. So these helpers could be applied only by people who really want them. That would potentially allow us to integrate the client types and have them only kick in if the provided game type allowed it.

If you fancy playing around with any of this, I’d be happy to take a look or answer questions.

**Reply by drmercer** [reply]
> One other stumbling point is you still need to assert `as const`:

As far as I can tell, `as const` isn't necessary here, unless your object includes arrays you want treated as tuples or primitives you want typed as literals. All it does for objects themselves is make the properties readonly. (See [the release notes for `as const`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions).) In my testing, `defineGame` works fine without it.

> Also this makes it tricky to type `G`

Agreed, that is an unfortunate downside. I tried typing `G` by specifying a return type on the `setup` function (see below), but it doesn't work unless I also specify that type on the `g` param in the move functions. Vue manages to solve a similar problem somehow (the `props` param to their `setup` function is typed automagically based on the `props` options field). I wonder how to accomplish the same thing here. :thinking:

```ts
interface MyState {
  myStateProp: string;
}

export const MyGame = declareGame({
  setup(): MyState {
    return {
      myStateProp: 'hi',
    };
  },
  minPlayers: 2,
  moves: {
    myMove(g: MyState, ctx, arg1: string) {
      console.log(g.myStateProp);
    },
  },
});
```

I'll play around with this with the game I'm building and see if I can figure anything out.

**Reply by delucis** [reply]
Cleaned up the example linked above so it works with strict type checks and incorporated the curried `defineGame` function. You can [try it out on TS Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQDRwKpwL5wGZQQhwBEMAzgLQwQQA2ARgKZ0wkDcAUKJLIp3DgBhGAA9UAuAHEAhiCYTBAGQgA7AOYAxaCACyEAG4LJ+o4rimmm1ecu6ZYcwAUAFjPJMhavMHXO3HvaOkgDKMDLqTEHmACoArlCqXqo+fpy4BESkDBAyUAAm6nJMAHTAEBycnAD0AFS1cACi4DAAnnAQDABWTADG8DRwcR5wwKqjqjBMUB795arkcADuLkzjMnAMwPAQeHBF8nC93r6ji3Gq+Uw+qkz5JXC11ZxtYExNLa3I3X3wALyIbBcGr1AQNZJGPgbECGd54C5zNRwQbbRb5YDkMAyGC9VZQdDkGBQYBgMBjdTI1bSOAAMiOYjgeXUcXkk3IJTBcAAIhisTi8YsZHQljJWosoEwYAlxgYIMB7mDnq93sgQNsAIJQdTkAA8mgAfHAAZo4ExRFNLosABRSABcjNUrXQ-VE9pE4jgJS9TPI9rGeGmcHVAEojYaZI7JAB+OBWr0lH32kNhuCy+WSe23SHAuq1TmNc1QGT9RlwGFGfAImDzfCEYg5GAuODkFzQGCUB35OB0NTqSh4HRl2FwK69Oh5bHzdmKl6tN4WWG6SWt-I6mKm81rfKLSyGgHrs0W7cLozWaNwGIZi8bo+LFQabRQPSw88xADaJHLTBIAF0r1nphzUFHiaQti3gaFh3RTFsVxQNMmIVFm1bWAO2gbte37QcvxHPpxyLas1A5R4lTnd5LB5GD+WmNcby3HdYT3FA1RgTVtR1OxlwgVcYn1fUgLzECIWmCCywcDo9hw0d8MnNRFkGaFxN2IcK2gvk4JmYinlnedLHIGIIEo9SBVow96JPKJxIAHyGS5rjGO4mIPTdLQsoJzwQSRBEEN8AGkJjgABrJhWmUmIf3tCjeVgvE1z8n9+K8nAr2aMA2m+Hp+gE-MwJLRSwAklT3jUmLpkWBDSxAJS9kJCJirwidCIWLTSPnMI6qM0qZlMlzj3ayIgjgGyLiuG5HKNTASgASUmMrfmQPAdUkZASgwVR5gWpbvOvMzXP6yywCSwQY087azrgPyAuC0K9nCyLYX0wzouo7r318n8Py-chf0S87BGwI64HtVL0p+fpJH1ThEpBQSGgLIlwMq6qipHZ6NPK9CNgOJhAEwCeSIDASg6CYIw6A6KBS3IclibgMAAlKGdlW5NGBU0OtklSHrbzgVx3E8E51CYvSDM6l7dUkd9Pwe38odpSQrUlqVEl-OjXPiRIOd8JKY320X0bishpV-D9asib6EqS4HPgy35g2ykD4aLPKxIK5SpJZsrayyfLCrpvncLHRqpxanT3l5jw9ZM5zufDg6hts0aHPyJiMGm2aZnmxbltW9a1E2pLo-M2P3POk7Ae2y6xiCkKwoi5mqPRtmiE19Q4ve36-oB86rbSr4wZgCGoftuHctEqrXckqCPZmL3iCx4oA5kpqQ6ZyOyqbkBZHkLnzK3pgmIQXA6TXmYN5btdDRpSRY5P3VJb9jxzeh3N6lft-34-z+v+-n-X85dUMAxAABLIAAEpTQAHJSDgEAxoSgnCNFAWCX+KDUFoK-s8GGnIhAJGJHcWm7gqCNkIHEdQTZ4SqEROMQY5A4ikjbMiMicAXBjAoJ2TkMg4g0EoMccAxMmrLG2K2LhGEqYaEpO8MAEpegYneNjRhbwtKcgAOpCIgCIxsGJ0ChTiEcdRdAuz5AgHAAABtje0e8dRSENBNJAXocAmM7KIpgnJGySIgOQKmWw6DbHaMpK4xIDDkiOD4tY8ASovQUUwdkKAKY6L0XEAxnJeh5GuIkug7QmYk2mK0TR4iGDtDcJcQRjY4CqGMZw7hvCwD8PeBGLsagMn7ElK4phcFeiBWCSsNYBDPHBLcfsBeid1oCMGDkPIhRihlAgFpOAIRjGaMWBiEJFxApOk2CIgIow2F0KplcacJFOAUKobhMaljoG7WPBGdoAJrn6itKGU6cAJRK3GMcgRVjiCXMWOcviVpzHSBAMGCxxAnmCBedKQZ8guD-S4ADGG6DEVIu-pyDAIR1RSEaMg5FOLcWYNYdMPAxZyKtGgU88gxwJSZhZMwKAcKqjHAWPAeRAJhlMEsboUl9zgxWieXUC8bTVgdPklSCFiRUxCjiO8RlA4nz41MVIRxgBeDcALI7hzBAeClGAe0DyUy8ubJSpg9oAAMOBgzmC+vaMF+BaA6qkM6MQ6AmTUpALS0MfxDTWsEPymIgq+iBRFdiUxTJHEpPGMwRkfT1C3C7IMExFLoBMGVWq7S50pAlATRKCaTIYXeWwOYfNkhXlWqSqbaJJbzrqgrX9S1-A-reQYHkHV7rPWFvrYIRtAAvO1DqPQgHIOoe0hJiQaAJH0F5UZ7Q5HoEwCMLbATmHOm2s6y7-oFokNgO2VRcwqOlRGMpEAlg3idgMKkX1Z4SKhfVG42waz1L0aoSEA8QJuOIIMNxwAKYRI0jap8sYlhqJEdAukLpGRahZGE8gwYV5MKEKEyYekJq3w3pY5UylsaJWqNUWMzqyk0umPOtM+RQ42uMQCODwAwl6Q-AOCoP4uBYdjIRuUxGmaNopuR+DMBqMkHY7+Bj2GrT9sHc2Ik5Ix29AnVO2gxN93DTsmNfIzH0xsZkJ2iaFGqMPQ-F2-jnAgA).
